File: c/src/matrix/basic_matrix_multiplier.c
#include "basic_matrix_multiplier.h"
#include <stdlib.h>

double **allocate_matrix(int n) {
    double **matrix = (double **)malloc(n * sizeof(double *));
    for (int i = 0; i < n; ++i)
        matrix[i] = (double *)malloc(n * sizeof(double));
    return matrix;
}

void free_matrix(double **matrix, int n) {
    for (int i = 0; i < n; ++i)
        free(matrix[i]);
    free(matrix);
}

void generate_matrices(double **A, double **B, int n) {
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            A[i][j] = (double)rand() / RAND_MAX;
            B[i][j] = (double)rand() / RAND_MAX;
        }
    }
}

void multiply(double **A, double **B, double **C, int n) {
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            C[i][j] = 0.0;
            for (int k = 0; k < n; ++k)
                C[i][j] += A[i][k] * B[k][j];
        }
    }
}

File: python/src/matrix/basic_matrix_multiplier.py
import random

class BasicMatrixMultiplier:
    def __init__(self):
        self.random = random.Random()

    def generate_matrices(self, n):
        A = [[self.random.random() for _ in range(n)] for _ in range(n)]
        B = [[self.random.random() for _ in range(n)] for _ in range(n)]
        return A, B

    def multiply(self, first_matrix, second_matrix):
        size = len(first_matrix)
        result_matrix = [[0.0 for _ in range(size)] for _ in range(size)]

        for i in range(size):
            for j in range(size):
                for k in range(size):
                    result_matrix[i][j] += first_matrix[i][k] * second_matrix[k][j]
        return result_matrix